File: crates/nw-gamedata-codegen/resources/typescript/standalone/game-assets/datasheet.ts
const DATASHEET_VERSIONS = [0x11, 0x12] as const;
const DATASHEET_NAME_CRC_OFFSET = 0x04;
const DATASHEET_NAME_STRING_OFFSET = 0x08;
const DATASHEET_TYPE_CRC_OFFSET = 0x0c;
const DATASHEET_TYPE_STRING_OFFSET = 0x10;
const DATASHEET_DATA_SIZE_OFFSET = 0x38;
const DATASHEET_COLUMN_COUNT_OFFSET = 0x44;
const DATASHEET_ROW_COUNT_OFFSET = 0x48;
const DATASHEET_COLUMN_RECORDS_OFFSET = 0x5c;
const DATASHEET_DATA_END_OFFSET = 0x38;
const DATASHEET_U32_SIZE = 4;
const DATASHEET_COLUMN_RECORD_SIZE = 12;
const DATASHEET_CELL_RECORD_SIZE = 8;

export enum DatasheetColumnType {
  String = 0x01,
  Number = 0x02,
  Boolean = 0x03
}

export interface Datasheet {
  readonly version: number;
  readonly nameCrc: number;
  readonly name: string;
  readonly typeCrc: number;
  readonly typeName: string;
  readonly columns: readonly DatasheetColumn[];
  readonly rows: readonly DatasheetRow[];
}

export interface DatasheetColumn {
  readonly crc: number;
  readonly name: string;
  readonly columnType: DatasheetColumnType;
}

export interface DatasheetRow {
  readonly cells: readonly DatasheetCell[];
}

export interface DatasheetCell {
  readonly crc: number;
  readonly value: DatasheetCellValue;
}

export type DatasheetCellValue = { readonly kind: "string"; readonly value: string } | { readonly kind: "number"; readonly value: number } | { readonly kind: "boolean"; readonly value: boolean };

export function parseDatasheet(bytes: Uint8Array): Datasheet {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const version = readU32(view, 0x00, "datasheet version");
  if (!DATASHEET_VERSIONS.includes(version as 0x11 | 0x12)) {
    throw new Error(`unsupported datasheet version 0x${version.toString(16)}`);
  }

  const nameCrc = readU32(view, DATASHEET_NAME_CRC_OFFSET, "datasheet name crc");
  const nameOffset = readU32(view, DATASHEET_NAME_STRING_OFFSET, "datasheet name string offset");
  const typeCrc = readU32(view, DATASHEET_TYPE_CRC_OFFSET, "datasheet type crc");
  const typeOffset = readU32(view, DATASHEET_TYPE_STRING_OFFSET, "datasheet type string offset");
  const dataSize = readU32(view, DATASHEET_DATA_SIZE_OFFSET, "datasheet data size");
  const columnCount = readU32(view, DATASHEET_COLUMN_COUNT_OFFSET, "datasheet column count");
  const rowCount = readU32(view, DATASHEET_ROW_COUNT_OFFSET, "datasheet row count");
  const columnsOffset = DATASHEET_COLUMN_RECORDS_OFFSET;
  const columnsLength = checkedProduct(columnCount, DATASHEET_COLUMN_RECORD_SIZE, "datasheet column records");
  const cellsOffset = columnsOffset + columnsLength;
  const cellCount = checkedProduct(rowCount, columnCount, "datasheet cell count");
  const cellsLength = checkedProduct(cellCount, DATASHEET_CELL_RECORD_SIZE, "datasheet cell records");
  const stringsOffset = cellsOffset + cellsLength;
  const expectedStringsOffset = DATASHEET_DATA_END_OFFSET + DATASHEET_U32_SIZE + dataSize;
  if (stringsOffset !== expectedStringsOffset) {
    throw new Error(`datasheet data_size points string table at 0x${expectedStringsOffset.toString(16)}, expected 0x${stringsOffset.toString(16)}`);
  }
  assertByteRange(bytes, stringsOffset, bytes.byteLength - stringsOffset, "datasheet string table");

  const stringAt = (offset: number, label: string): string => readString(bytes, stringsOffset, offset, label);
  const columns: DatasheetColumn[] = [];
  for (let columnIndex = 0; columnIndex < columnCount; columnIndex += 1) {
    const offset = columnsOffset + columnIndex * DATASHEET_COLUMN_RECORD_SIZE;
    assertByteRange(bytes, offset, DATASHEET_COLUMN_RECORD_SIZE, `datasheet column ${columnIndex}`);
    const rawColumnType = readU32(view, offset + 8, `datasheet column ${columnIndex} type`);
    const columnType = parseColumnType(rawColumnType, columnIndex);
    columns.push({
      crc: readU32(view, offset, `datasheet column ${columnIndex} crc`),
      name: stringAt(readI32(view, offset + 4, `datasheet column ${columnIndex} name offset`), `datasheet column ${columnIndex} name`),
      columnType
    });
  }

  const rows: DatasheetRow[] = [];
  for (let rowIndex = 0; rowIndex < rowCount; rowIndex += 1) {
    const cells: DatasheetCell[] = [];
    for (let columnIndex = 0; columnIndex < columnCount; columnIndex += 1) {
      const record = cellsOffset + (rowIndex * columnCount + columnIndex) * DATASHEET_CELL_RECORD_SIZE;
      assertByteRange(bytes, record, DATASHEET_CELL_RECORD_SIZE, `datasheet row ${rowIndex} column ${columnIndex}`);
      cells.push({
        crc: readU32(view, record, `datasheet row ${rowIndex} column ${columnIndex} crc`),
        value: parseCellValue(view, stringAt, record + 4, columns[columnIndex] as DatasheetColumn, rowIndex)
      });
    }
    rows.push({ cells });
  }

  return {
    version,
    nameCrc,
    name: stringAt(nameOffset, "datasheet name"),
    typeCrc,
    typeName: stringAt(typeOffset, "datasheet type name"),
    columns,
    rows
  };
}

function parseColumnType(raw: number, columnIndex: number): DatasheetColumnType {
  switch (raw) {
    case DatasheetColumnType.String:
      return DatasheetColumnType.String;
    case DatasheetColumnType.Number:
      return DatasheetColumnType.Number;
    case DatasheetColumnType.Boolean:
      return DatasheetColumnType.Boolean;
    default:
      throw new Error(`unknown datasheet column type ${raw} at column ${columnIndex}`);
  }
}

function parseCellValue(view: DataView, stringAt: (offset: number, label: string) => string, valueOffset: number, column: DatasheetColumn, rowIndex: number): DatasheetCellValue {
  switch (column.columnType) {
    case DatasheetColumnType.String:
      return {
        kind: "string",
        value: stringAt(readU32(view, valueOffset, `${column.name} string offset`), `row ${rowIndex} ${column.name}`)
      };
    case DatasheetColumnType.Number:
      return { kind: "number", value: view.getFloat32(valueOffset, true) };
    case DatasheetColumnType.Boolean:
      return { kind: "boolean", value: view.getInt32(valueOffset, true) !== 0 };
  }
}

function readU32(view: DataView, offset: number, label: string): number {
  assertViewRange(view, offset, 4, label);
  return view.getUint32(offset, true);
}

function readI32(view: DataView, offset: number, label: string): number {
  assertViewRange(view, offset, 4, label);
  return view.getInt32(offset, true);
}

function readString(bytes: Uint8Array, stringsOffset: number, offset: number, label: string): string {
  if (offset < 0) {
    throw new Error(`${label} has negative string offset ${offset}`);
  }
  const start = stringsOffset + offset;
  assertByteRange(bytes, start, 1, label);
  let end = start;
  while (end < bytes.byteLength && bytes[end] !== 0) {
    end += 1;
  }
  if (end === bytes.byteLength) {
    throw new Error(`${label} is unterminated`);
  }
  return new TextDecoder().decode(bytes.subarray(start, end));
}

function checkedProduct(left: number, right: number, label: string): number {
  const product = left * right;
  if (!Number.isSafeInteger(product)) {
    throw new Error(`${label} length is not a safe integer`);
  }
  return product;
}

function assertViewRange(view: DataView, offset: number, length: number, label: string): void {
  assertLength(view.byteLength, offset, length, label);
}

function assertByteRange(bytes: Uint8Array, offset: number, length: number, label: string): void {
  assertLength(bytes.byteLength, offset, length, label);
}

function assertLength(byteLength: number, offset: number, length: number, label: string): void {
  if (!Number.isSafeInteger(offset) || !Number.isSafeInteger(length) || offset < 0 || length < 0 || offset + length > byteLength) {
    throw new Error(`${label} out of bounds`);
  }
}

File: crates/nw-gamedata-codegen/resources/go/standalone/game-assets/datasheet.go
package gameassets

import (
	"encoding/binary"
	"fmt"
	"math"
)

const (
	datasheetNameCRCOffset       = 0x04
	datasheetNameStringOffset    = 0x08
	datasheetTypeCRCOffset       = 0x0c
	datasheetTypeStringOffset    = 0x10
	datasheetDataSizeOffset      = 0x38
	datasheetColumnCountOffset   = 0x44
	datasheetRowCountOffset      = 0x48
	datasheetColumnRecordsOffset = 0x5c
	datasheetDataEndOffset       = 0x38
	datasheetU32Size             = 4
	datasheetColumnRecordSize    = 12
	datasheetCellRecordSize      = 8
)

var datasheetVersions = map[uint32]struct{}{
	0x11: {},
	0x12: {},
}

type DatasheetColumnType uint32

const (
	DatasheetColumnString  DatasheetColumnType = 0x01
	DatasheetColumnNumber  DatasheetColumnType = 0x02
	DatasheetColumnBoolean DatasheetColumnType = 0x03
)

type Datasheet struct {
	Version  uint32
	NameCRC  uint32
	Name     string
	TypeCRC  uint32
	TypeName string
	Columns  []DatasheetColumn
	Rows     []DatasheetRow
}

type DatasheetColumn struct {
	CRC        uint32
	Name       string
	ColumnType DatasheetColumnType
}

type DatasheetRow struct {
	Cells []DatasheetCell
}

type DatasheetCell struct {
	CRC   uint32
	Value DatasheetCellValue
}

type DatasheetCellKind uint8

const (
	DatasheetCellString DatasheetCellKind = iota + 1
	DatasheetCellNumber
	DatasheetCellBoolean
)

type DatasheetCellValue struct {
	Kind    DatasheetCellKind
	String  string
	Number  float32
	Boolean bool
}

func ParseDatasheet(bytes []byte) (Datasheet, error) {
	version, err := readDatasheetU32(bytes, 0x00, "datasheet version")
	if err != nil {
		return Datasheet{}, err
	}
	if _, ok := datasheetVersions[version]; !ok {
		return Datasheet{}, fmt.Errorf("unsupported datasheet version %#x", version)
	}

	nameCRC, err := readDatasheetU32(bytes, datasheetNameCRCOffset, "datasheet name crc")
	if err != nil {
		return Datasheet{}, err
	}
	nameOffset, err := readDatasheetU32(bytes, datasheetNameStringOffset, "datasheet name string offset")
	if err != nil {
		return Datasheet{}, err
	}
	typeCRC, err := readDatasheetU32(bytes, datasheetTypeCRCOffset, "datasheet type crc")
	if err != nil {
		return Datasheet{}, err
	}
	typeOffset, err := readDatasheetU32(bytes, datasheetTypeStringOffset, "datasheet type string offset")
	if err != nil {
		return Datasheet{}, err
	}
	dataSize, err := readDatasheetU32(bytes, datasheetDataSizeOffset, "datasheet data size")
	if err != nil {
		return Datasheet{}, err
	}
	columnCount, err := readDatasheetU32(bytes, datasheetColumnCountOffset, "datasheet column count")
	if err != nil {
		return Datasheet{}, err
	}
	rowCount, err := readDatasheetU32(bytes, datasheetRowCountOffset, "datasheet row count")
	if err != nil {
		return Datasheet{}, err
	}

	columnRecordsLength, err := checkedDatasheetProduct(columnCount, datasheetColumnRecordSize, "datasheet column records")
	if err != nil {
		return Datasheet{}, err
	}
	cellCount, err := checkedDatasheetProduct(rowCount, columnCount, "datasheet cell count")
	if err != nil {
		return Datasheet{}, err
	}
	cellRecordsLength, err := checkedDatasheetProduct(cellCount, datasheetCellRecordSize, "datasheet cell records")
	if err != nil {
		return Datasheet{}, err
	}
	columnsOffset := datasheetColumnRecordsOffset
	cellsOffset := columnsOffset + int(columnRecordsLength)
	stringsOffset := cellsOffset + int(cellRecordsLength)
	expectedStringsOffset := datasheetDataEndOffset + datasheetU32Size + int(dataSize)
	if stringsOffset != expectedStringsOffset {
		return Datasheet{}, fmt.Errorf("datasheet data_size points string table at %#x, expected %#x", expectedStringsOffset, stringsOffset)
	}
	if err := requireDatasheetRange(bytes, stringsOffset, len(bytes)-stringsOffset, "datasheet string table"); err != nil {
		return Datasheet{}, err
	}

	stringAt := func(offset uint32, label string) (string, error) {
		return readDatasheetString(bytes, stringsOffset, int(offset), label)
	}
	name, err := stringAt(nameOffset, "datasheet name")
	if err != nil {
		return Datasheet{}, err
	}
	typeName, err := stringAt(typeOffset, "datasheet type name")
	if err != nil {
		return Datasheet{}, err
	}

	columns := make([]DatasheetColumn, 0, columnCount)
	for columnIndex := uint32(0); columnIndex < columnCount; columnIndex++ {
		record := columnsOffset + int(columnIndex)*datasheetColumnRecordSize
		if err := requireDatasheetRange(bytes, record, datasheetColumnRecordSize, fmt.Sprintf("datasheet column %d", columnIndex)); err != nil {
			return Datasheet{}, err
		}
		crc := binary.LittleEndian.Uint32(bytes[record : record+4])
		nameOffset := int32(binary.LittleEndian.Uint32(bytes[record+4 : record+8]))
		if nameOffset < 0 {
			return Datasheet{}, fmt.Errorf("datasheet column %d has negative string offset %d", columnIndex, nameOffset)
		}
		name, err := readDatasheetString(bytes, stringsOffset, int(nameOffset), fmt.Sprintf("datasheet column %d name", columnIndex))
		if err != nil {
			return Datasheet{}, err
		}
		columnType, err := parseDatasheetColumnType(binary.LittleEndian.Uint32(bytes[record+8:record+12]), columnIndex)
		if err != nil {
			return Datasheet{}, err
		}
		columns = append(columns, DatasheetColumn{
			CRC:        crc,
			Name:       name,
			ColumnType: columnType,
		})
	}

	rows := make([]DatasheetRow, 0, rowCount)
	for rowIndex := uint32(0); rowIndex < rowCount; rowIndex++ {
		cells := make([]DatasheetCell, 0, columnCount)
		for columnIndex := uint32(0); columnIndex < columnCount; columnIndex++ {
			record := cellsOffset + int(rowIndex*columnCount+columnIndex)*datasheetCellRecordSize
			if err := requireDatasheetRange(bytes, record, datasheetCellRecordSize, fmt.Sprintf("datasheet row %d column %d", rowIndex, columnIndex)); err != nil {
				return Datasheet{}, err
			}
			value, err := parseDatasheetCellValue(bytes, stringsOffset, record+4, columns[columnIndex], rowIndex)
			if err != nil {
				return Datasheet{}, err
			}
			cells = append(cells, DatasheetCell{
				CRC:   binary.LittleEndian.Uint32(bytes[record : record+4]),
				Value: value,
			})
		}
		rows = append(rows, DatasheetRow{Cells: cells})
	}

	return Datasheet{
		Version:  version,
		NameCRC:  nameCRC,
		Name:     name,
		TypeCRC:  typeCRC,
		TypeName: typeName,
		Columns:  columns,
		Rows:     rows,
	}, nil
}

func parseDatasheetColumnType(raw uint32, columnIndex uint32) (DatasheetColumnType, error) {
	switch DatasheetColumnType(raw) {
	case DatasheetColumnString:
		return DatasheetColumnString, nil
	case DatasheetColumnNumber:
		return DatasheetColumnNumber, nil
	case DatasheetColumnBoolean:
		return DatasheetColumnBoolean, nil
	default:
		return 0, fmt.Errorf("unknown datasheet column type %d at column %d", raw, columnIndex)
	}
}

func parseDatasheetCellValue(bytes []byte, stringsOffset int, valueOffset int, column DatasheetColumn, rowIndex uint32) (DatasheetCellValue, error) {
	switch column.ColumnType {
	case DatasheetColumnString:
		offset, err := readDatasheetU32(bytes, valueOffset, column.Name+" string offset")
		if err != nil {
			return DatasheetCellValue{}, err
		}
		value, err := readDatasheetString(bytes, stringsOffset, int(offset), fmt.Sprintf("row %d %s", rowIndex, column.Name))
		if err != nil {
			return DatasheetCellValue{}, err
		}
		return DatasheetCellValue{Kind: DatasheetCellString, String: value}, nil
	case DatasheetColumnNumber:
		if err := requireDatasheetRange(bytes, valueOffset, 4, column.Name+" number"); err != nil {
			return DatasheetCellValue{}, err
		}
		return DatasheetCellValue{
			Kind:   DatasheetCellNumber,
			Number: math.Float32frombits(binary.LittleEndian.Uint32(bytes[valueOffset : valueOffset+4])),
		}, nil
	case DatasheetColumnBoolean:
		if err := requireDatasheetRange(bytes, valueOffset, 4, column.Name+" boolean"); err != nil {
			return DatasheetCellValue{}, err
		}
		return DatasheetCellValue{
			Kind:    DatasheetCellBoolean,
			Boolean: int32(binary.LittleEndian.Uint32(bytes[valueOffset:valueOffset+4])) != 0,
		}, nil
	default:
		return DatasheetCellValue{}, fmt.Errorf("unknown datasheet column type %d for %s", column.ColumnType, column.Name)
	}
}

func readDatasheetU32(bytes []byte, offset int, label string) (uint32, error) {
	if err := requireDatasheetRange(bytes, offset, 4, label); err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint32(bytes[offset : offset+4]), nil
}

func readDatasheetString(bytes []byte, stringsOffset int, offset int, label string) (string, error) {
	if offset < 0 {
		return "", fmt.Errorf("%s has negative string offset %d", label, offset)
	}
	start := stringsOffset + offset
	if err := requireDatasheetRange(bytes, start, 1, label); err != nil {
		return "", err
	}
	end := start
	for end < len(bytes) && bytes[end] != 0 {
		end++
	}
	if end == len(bytes) {
		return "", fmt.Errorf("%s is unterminated", label)
	}
	return string(bytes[start:end]), nil
}

func checkedDatasheetProduct(left uint32, right uint32, label string) (uint32, error) {
	product := uint64(left) * uint64(right)
	if product > 0xffff_ffff {
		return 0, fmt.Errorf("%s length overflow", label)
	}
	return uint32(product), nil
}

func requireDatasheetRange(bytes []byte, offset int, length int, label string) error {
	if offset < 0 || length < 0 || offset+length < offset || offset+length > len(bytes) {
		return fmt.Errorf("%s out of bounds", label)
	}
	return nil
}
